// badgeforcejs/lib/src/contract/contract-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Web3 = require("web3");
var contract = require("truffle-contract");
var ContractConfig = /** @class */ (function () {
    /**
     * [constructor description]
     * @param web3Provider     web3 provider
     * @param artifactLocation JSON ABI of the contract
     * @param accountIndex     index of the account to use for sending transactions
     * @param gas              default gas
     */
    function ContractConfig(web3Provider, artifactLocation, accountIndex, gas) {
        this.web3Provider = web3Provider;
        this.artifactLocation = artifactLocation;
        this.accountIndex = accountIndex;
        this.gas = gas;
    }
    return ContractConfig;
}());
exports.ContractConfig = ContractConfig;
var Contract = /** @class */ (function () {
    /**
     * @param config configuration object [[ContractConfig]]
     */
    function Contract(config) {
        this.config = config;
        this.web3 = new Web3(new Web3.providers.HttpProvider(config.web3Provider));
        this.contractABI = contract(config.artifactLocation);
        this.contractABI.setProvider(this.web3.currentProvider);
        this.contractABI.defaults({
            from: this.web3.eth.accounts[config.accountIndex],
            gas: config.gas,
        });
    }
    return Contract;
}());
exports.Contract = Contract;
