// badgeforcejs/lib/src/issuer/api.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var contract_1 = require("../contract");
var obj_helpers_1 = require("./obj_helpers");
var Issuer = /** @class */ (function () {
    /**
     * @param issuer instance from truffle-contract
     */
    function Issuer(issuer) {
        this.issuer = issuer;
    }
    /**
     * Gets the public information for this issuer contract
     * @return info object
     */
    Issuer.prototype.getIssuerInfo = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, _c, _d, _e;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        _b = (_a = obj_helpers_1.IssuerInfo.bind).apply;
                        _c = [obj_helpers_1.IssuerInfo];
                        _e = (_d = [void 0]).concat;
                        return [4 /*yield*/, this.issuer.getInfo()];
                    case 1: return [2 /*return*/, new (_b.apply(_a, _c.concat([_e.apply(_d, [(_f.sent())])])))()];
                }
            });
        });
    };
    /**
     * Gets a badge by index (index of the badge name on the contract)
     * @param  i [description]
     * @return   badge object
     */
    Issuer.prototype.getBadge = function (i) {
        return __awaiter(this, void 0, void 0, function () {
            var name_1, _a, _b, _c, _d, _e, error_1;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        _f.trys.push([0, 3, , 4]);
                        return [4 /*yield*/, this.issuer.getNameByIndex(i)];
                    case 1:
                        name_1 = _f.sent();
                        _b = (_a = obj_helpers_1.IssuerBadge.bind).apply;
                        _c = [obj_helpers_1.IssuerBadge];
                        _e = (_d = [void 0]).concat;
                        return [4 /*yield*/, this.issuer.getBadge(name_1)];
                    case 2: return [2 /*return*/, new (_b.apply(_a, _c.concat([_e.apply(_d, [(_f.sent())])])))()];
                    case 3:
                        error_1 = _f.sent();
                        throw new Error(error_1);
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Gets all the badges
     * @return array of badges
     */
    Issuer.prototype.getBadges = function () {
        return __awaiter(this, void 0, void 0, function () {
            var badges, count, i, _a, _b, error_2;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 6, , 7]);
                        badges = [];
                        return [4 /*yield*/, this.issuer.getNumberOfBadges()];
                    case 1:
                        count = _c.sent();
                        i = 0;
                        _c.label = 2;
                    case 2:
                        if (!(i < count.toString())) return [3 /*break*/, 5];
                        _b = (_a = badges).push;
                        return [4 /*yield*/, this.getBadge(i)];
                    case 3:
                        _b.apply(_a, [_c.sent()]);
                        _c.label = 4;
                    case 4:
                        i++;
                        return [3 /*break*/, 2];
                    case 5: return [2 /*return*/, badges];
                    case 6:
                        error_2 = _c.sent();
                        throw new Error(error_2);
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Creates a new badge for issuance
     * @param  badge badge object
     */
    Issuer.prototype.createBadge = function (badge) {
        return __awaiter(this, void 0, void 0, function () {
            var error_3, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, (_a = this.issuer).createBadge.apply(_a, Object.values(badge))];
                    case 1: return [2 /*return*/, _b.sent()];
                    case 2:
                        error_3 = _b.sent();
                        throw new Error(error_3);
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Deletes a badge by name
     * @param  name name of the badge to be deleted
     */
    Issuer.prototype.deleteBadge = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            var error_4;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.issuer.deleteBadge(name)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        error_4 = _a.sent();
                        throw new Error(error_4);
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Issues a credential
     * @param  credential credential object
     */
    Issuer.prototype.issueCredential = function (credential) {
        return __awaiter(this, void 0, void 0, function () {
            var error_5, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, (_a = this.issuer).issue.apply(_a, Object.values(credential))];
                    case 1: return [2 /*return*/, _b.sent()];
                    case 2:
                        error_5 = _b.sent();
                        throw new Error(error_5);
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Gets current number of badges on the contract
     * @return number of badges
     */
    Issuer.prototype.getNumberOfBadges = function () {
        return __awaiter(this, void 0, void 0, function () {
            var error_6;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.issuer.getNumberOfBadges()];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        error_6 = _a.sent();
                        throw new Error(error_6);
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Authorizes a new ethereum wallet address. This will allow an address to do things.
     * @param  newIssuer wallet address
     */
    Issuer.prototype.authorzeAccount = function (newIssuer) {
        return __awaiter(this, void 0, void 0, function () {
            var error_7;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.issuer.authorzeAccount(newIssuer)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        error_7 = _a.sent();
                        throw new Error(error_7);
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Removes an authorized address. Once this is done, the removed address will no longer be able to interact with the contract
     * @param  issuer wallet address
     */
    Issuer.prototype.removeAuthorizedAccount = function (issuer) {
        return __awaiter(this, void 0, void 0, function () {
            var error_8;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.issuer.removeAuthorizedAccount(issuer)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        error_8 = _a.sent();
                        throw new Error(error_8);
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Gets an authorized account (wallet address) by index
     * @param  index wallet address
     * @return       address of the authorized account
     */
    Issuer.prototype.getAuthorizedAccount = function (index) {
        return __awaiter(this, void 0, void 0, function () {
            var error_9;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.issuer.getAuthorizedAccount(index)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        error_9 = _a.sent();
                        throw new Error(error_9);
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Gets the current number of authorized accounts on the contract
     * @return number of authorized accounts
     */
    Issuer.prototype.getNumberOfAuthorizedAccounts = function () {
        return __awaiter(this, void 0, void 0, function () {
            var error_10;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.issuer.getNumberOfAuthorizedAccounts()];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        error_10 = _a.sent();
                        throw new Error(error_10);
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Revoke a credential by transaction key
     * @param  txnKey transaction key of the credential
     */
    Issuer.prototype.revoke = function (txnKey) {
        return __awaiter(this, void 0, void 0, function () {
            var error_11;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.issuer.revoke(txnKey)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        error_11 = _a.sent();
                        throw new Error(error_11);
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Un-Revoke a credential by transaction key
     * @param  txnKey transaction key of the credential
     */
    Issuer.prototype.unRevoke = function (txnKey) {
        return __awaiter(this, void 0, void 0, function () {
            var error_12;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.issuer.unRevoke(txnKey)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        error_12 = _a.sent();
                        throw new Error(error_12);
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Checks if a credential is revoked by transaction key
     * @param  txnKey transaction key of the credential
     * @return true/false
     */
    Issuer.prototype.isRevoked = function (txnKey) {
        return __awaiter(this, void 0, void 0, function () {
            var error_13;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.issuer.isRevoked(txnKey)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        error_13 = _a.sent();
                        throw new Error(error_13);
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Verifies a credential
     * @param  txnKey    transaction key of the credential
     * @param  recipient recipient of the credential (holder contract address)
     * @return           verification response
     */
    Issuer.prototype.verifyCredential = function (txnKey, recipient) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, _c, _d, _e, error_14;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        _f.trys.push([0, 2, , 3]);
                        _b = (_a = obj_helpers_1.VerifyResponse.bind).apply;
                        _c = [obj_helpers_1.VerifyResponse];
                        _e = (_d = [void 0]).concat;
                        return [4 /*yield*/, this.issuer.verifyCredential(txnKey, recipient)];
                    case 1: return [2 /*return*/, new (_b.apply(_a, _c.concat([_e.apply(_d, [_f.sent()])])))()];
                    case 2:
                        error_14 = _f.sent();
                        throw new Error(error_14);
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    return Issuer;
}());
exports.Issuer = Issuer;
var IssuerClient = /** @class */ (function (_super) {
    __extends(IssuerClient, _super);
    /**
     * @param config a configuration object
     */
    function IssuerClient(config) {
        return _super.call(this, config) || this;
    }
    /**
     * Gets a [[Issuer]] object instanstiated with contract data at the given address
     * @param  address address of the contract
     * @return         a issuer object
     */
    IssuerClient.prototype.getInstance = function (address) {
        return __awaiter(this, void 0, void 0, function () {
            var instance, error_15;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.contractABI.at(address)];
                    case 1:
                        instance = _a.sent();
                        return [2 /*return*/, new Issuer(instance)];
                    case 2:
                        error_15 = _a.sent();
                        throw new Error(error_15);
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    return IssuerClient;
}(contract_1.Contract));
exports.IssuerClient = IssuerClient;
