// badgeforcejs/lib/src/issuer/obj_helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var IssuerInfo = /** @class */ (function () {
    /**
     * Issuer info helper object
     * @param issuer   address of the issuer god account (wallet address)
     * @param contract address of the issuer contract
     * @param name      name of the issuer
     * @param url      link to issuer site or page
     */
    function IssuerInfo(issuer, contract, name, url) {
        this.issuer = issuer;
        this.contract = contract;
        this.name = name;
        this.url = url;
    }
    return IssuerInfo;
}());
exports.IssuerInfo = IssuerInfo;
var IssuerBadge = /** @class */ (function () {
    /**
     * Issuer badge helper object
     * @param address     address of the issuer contract
     * @param description
     * @param name
     * @param image
     * @param version
     */
    function IssuerBadge(address, description, name, image, version) {
        this.address = address;
        this.description = description;
        this.name = name;
        this.image = image;
        this.version = version;
    }
    return IssuerBadge;
}());
exports.IssuerBadge = IssuerBadge;
var Credential = /** @class */ (function () {
    /**
     * Credential helper object. Used to mitigate errors when ussing [[issueCredential]] function
     * @param badgeName name of the badge that will be used for issuance
     * @param recipient address of the holder contract
     * @param expires   optional expiration date (pass 0 for never)
     */
    function Credential(badgeName, recipient, expires) {
        this.badgeName = badgeName;
        this.recipient = recipient;
        this.expires = expires;
    }
    return Credential;
}());
exports.Credential = Credential;
var VerifyResponse = /** @class */ (function () {
    /**
     * Verification response helper object
     * @param verified true/false
     * @param message  success or error message
     */
    function VerifyResponse(verified, message) {
        this.verified = verified;
        this.message = message;
    }
    return VerifyResponse;
}());
exports.VerifyResponse = VerifyResponse;
