// badgeforcejs/lib/src/factories/holder-factory.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var contract_1 = require("../contract");
/**
 * Parses response from EVM for new contract creation
 * @param holderAddrRes      response from EVM
 * @return the address of the new contract
 */
exports.parseHolderAddrRes = function (holderAddrRes) {
    return holderAddrRes.logs[0].args._location;
};
var HolderFactory = /** @class */ (function () {
    /**
     * @param factory      instance from truffle-contract
     * @param holderClient instanstiated [[HolderClient]] object. This is used so that we can get a nice usable instance of the new contract right away
     */
    function HolderFactory(factory, holderClient) {
        this.factory = factory;
        this.holderClient = holderClient;
    }
    /**
     * Creates a new holder contract
     * @param  owner wallet address of the real world recipient
     * @return       an instanstiated [[Holder]] object
     */
    HolderFactory.prototype.newHolder = function (owner) {
        return __awaiter(this, void 0, void 0, function () {
            var newHolderAddr, address, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.factory.createHolderContract(owner)];
                    case 1:
                        newHolderAddr = _a.sent();
                        address = exports.parseHolderAddrRes(newHolderAddr);
                        return [2 /*return*/, this.holderClient.getInstance(address)];
                    case 2:
                        error_1 = _a.sent();
                        throw new Error(error_1);
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    return HolderFactory;
}());
exports.HolderFactory = HolderFactory;
var HolderFactoryClient = /** @class */ (function (_super) {
    __extends(HolderFactoryClient, _super);
    /**
     * @param config       a configuration object [[ContractConfig]]
     * @param holderClient instanstiated [[HolderClient]] object. This is used so that we can get a nice usable instance of the new contract right away
     */
    function HolderFactoryClient(config, holderClient) {
        var _this = _super.call(this, config) || this;
        _this.holderClient = holderClient;
        return _this;
    }
    /**
     * @param  address address of the contract
     * @return         an instanstiated [[HolderFactory]] object
     */
    HolderFactoryClient.prototype.getInstance = function (address) {
        return __awaiter(this, void 0, void 0, function () {
            var instance, error_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        return [4 /*yield*/, this.contractABI.at(address)];
                    case 1:
                        instance = _a.sent();
                        return [4 /*yield*/, new HolderFactory(instance, this.holderClient)];
                    case 2: return [2 /*return*/, _a.sent()];
                    case 3:
                        error_2 = _a.sent();
                        return [2 /*return*/, error_2];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    return HolderFactoryClient;
}(contract_1.Contract));
exports.HolderFactoryClient = HolderFactoryClient;
