// badgeforcejs/lib/src/holder/obj_helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var HolderInfo = /** @class */ (function () {
    /**
     * Holder info object helper
     * @param address address of the holder
     */
    function HolderInfo(address) {
        this.address = address;
    }
    return HolderInfo;
}());
exports.HolderInfo = HolderInfo;
var Credential = /** @class */ (function () {
    /**
     * Credential object helper
     * @param issuer      address of the issuer who issued this credential
     * @param description
     * @param name
     * @param image
     * @param version
     * @param expires     expiration date
     * @param recipient   address of the holder contract
     * @param txKey       transaction key of the credential
     * @param active      flag for pending credentials
     */
    function Credential(issuer, description, name, image, version, expires, recipient, txKey, active) {
        this.issuer = issuer;
        this.description = description;
        this.name = name;
        this.image = image;
        this.version = version;
        this.expires = expires;
        this.recipient = recipient;
        this.txKey = txKey;
        this.active = active;
    }
    return Credential;
}());
exports.Credential = Credential;
