// badgeforcejs/src/contract/contract-api.ts
import * as Web3 from 'web3';
import * as contract from 'truffle-contract';

export class ContractConfig {
  /**
   * [constructor description]
   * @param web3Provider     web3 provider
   * @param artifactLocation JSON ABI of the contract
   * @param accountIndex     index of the account to use for sending transactions
   * @param gas              default gas
   */
    constructor(public web3Provider: string, public artifactLocation: string, public accountIndex: number, public gas: number) {}
}

export class Contract {
	public contractABI: any;
	public web3: any;
  /**
   * @param config configuration object [[ContractConfig]]
   */
	constructor(public config: ContractConfig) {
	  this.web3 = new Web3(new Web3.providers.HttpProvider(config.web3Provider));
    this.contractABI = contract(config.artifactLocation);
    this.contractABI.setProvider(this.web3.currentProvider);
    this.contractABI.defaults({
        from: this.web3.eth.accounts[config.accountIndex],
        gas: config.gas,
    });
  }
}
